kani-compiler/src/kani_middle/attributes.rs

```diff
@@ -43,60 +43,170 @@ impl KaniAttributeKind {
+    }
+
+    /// Check that all attributes assigned to an item is valid.
+    /// Errors will be added to the session. Invoke self.tcx.sess.abort_if_errors() to terminate
+    /// the session and emit all errors found.
+    pub(super) fn check_attributes(&self) {
+        // Check that all attributes are correctly used and well formed.
+        let is_harness = self.is_harness();
+        for (&kind, attrs) in self.map.iter() {
+            if !is_harness && kind.is_harness_only() {
+                self.tcx.sess.span_err(
+                    attrs[0].span,
+                    format!(
+                        "the `{}` attribute also requires the `#[kani::proof]` attribute",
+                        kind.as_ref()
+                    ),
+                );
             }
-            KaniAttributeKind::Unwind => {
-                expect_single(tcx, kind, &attrs);
-                attrs.iter().for_each(|attr| {
-                    parse_unwind(tcx, attr);
-                })
+            match kind {
+                KaniAttributeKind::ShouldPanic => {
+                    expect_single(self.tcx, kind, &attrs);
+                    attrs.iter().for_each(|attr| {
+                        expect_no_args(self.tcx, kind, attr);
+                    })
+                }
+                KaniAttributeKind::Solver => {
+                    expect_single(self.tcx, kind, &attrs);
+                    attrs.iter().for_each(|attr| {
+                        parse_solver(self.tcx, attr);
+                    })
+                }
+                KaniAttributeKind::Stub => {
+                    parse_stubs(self.tcx, self.item, attrs);
+                }
+                KaniAttributeKind::Unwind => {
+                    expect_single(self.tcx, kind, &attrs);
+                    attrs.iter().for_each(|attr| {
+                        parse_unwind(self.tcx, attr);
+                    })
+                }
+                KaniAttributeKind::Proof => {
+                    expect_single(self.tcx, kind, &attrs);
+                    attrs.iter().for_each(|attr| check_proof_attribute(self.tcx, self.item, attr))
+                }
+                KaniAttributeKind::Unstable => attrs.iter().for_each(|attr| {
+                    let _ = UnstableAttribute::try_from(*attr).map_err(|err| err.report(self.tcx));
+                }),
             }
-            KaniAttributeKind::Proof => {
-                expect_single(tcx, kind, &attrs);
-                attrs.iter().for_each(|attr| check_proof_attribute(tcx, def_id, attr))
+        }
+    }
+
+    /// Check that any unstable API has been enabled. Otherwise, emit an error.
+    ///
+    /// TODO: Improve error message by printing the span of the harness instead of the definition.
+    pub fn check_unstable_features(&self, enabled_features: &[String]) {
+        if !matches!(self.tcx.type_of(self.item).skip_binder().kind(), TyKind::FnDef(..)) {
+            // skip closures due to an issue with rustc.
```

Comment:
Nitpick: maybe we could add a warning here instead of just returning with no message. No if that would cause too much noise for the user.

Reply:
Too much noise. We could improve the warning and say that today you can't really attach a unstable attribute to a closure anyways.

Reply:
Hmm, I could have `KaniAttributes` store the span to the item it is on, that way I could also report that span here?

Reply:
Sorry, improve the comment

Reply:
I'm confused what do you mean?

Reply:
Something like this:
```suggestion
            // Skip closures since it shouldn't be possible to add an unstable attribute to them.
            // We have to explicitly skip them though due to an issue with rustc:
```